### PR TITLE
[Impeller] Incorporate filters in subpass coverage.

### PR DESCRIPTION
**Pull request by bdero**
Resolves https://github.com/flutter/flutter/issues/131182.

I adjusted the size of the matrix filtered circle in both SaveLayer Matrix filter test variants (post-filter and backdrop filter) to move the transformed circle outside the coverage area that would be computed if subpass filters aren't being accounted for. Further, we need to avoid intersecting with the pass coverage hint when the transformation being applied by the paint's filter graph causes the Entity's space basis to deviate from the parent pass basis. We're still able to apply it for simple translations, though.

Before:

Screenshot 2023-09-13 at 10 46 13 AM

After:

Screenshot 2023-09-13 at 10 48 38 AM

### CODE REVIEW

File: impeller/entity/contents/filters/filter_contents.h [hunk 1]
```diff
@@ -125,10 +125,22 @@ class FilterContents : public Contents {
...
       const std::string& label = "Filter Snapshot") const override;
 
+  // |Contents|
+  const FilterContents* AsFilter() const override;
+
   virtual Matrix GetLocalTransform(const Matrix& parent_transform) const;
 
   Matrix GetTransform(const Matrix& parent_transform) const;
 
+  /// @brief  Returns true of this filter graph performs basis transformations
```
Comment (jonahwilliams) [inline impeller/entity/contents/filters/filter_contents.h:135]:
```suggestion
  /// @brief  Returns true if this filter graph performs basis transformations
```

File: impeller/entity/contents/filters/filter_contents.h [hunk 2]
```diff
@@ -125,10 +125,22 @@ class FilterContents : public Contents {
...
+  ///         to the filtered content. For example: Rotating, scaling, and
+  ///         skewing are all basis transformations, but translating is not.
```
Comment (jonahwilliams) [inline impeller/entity/contents/filters/filter_contents.h:137]: Is translating the only kind of non-basis transform? Could we rephrase this as "TranslateOnly"?

File: impeller/entity/entity_pass.cc [hunk 3]
```diff
@@ -110,12 +110,39 @@ std::optional<Rect> EntityPass::GetElementsCoverage(
...
+        subpass_entity.SetTransformation(subpass.xformation_);
+        coverage = image_filter->GetCoverage(subpass_entity);
+      } else {
+        coverage = unfiltered_coverage;
+      }
+
+      if (coverage.has_value() && coverage_limit.has_value() &&
+          (!image_filter || !image_filter->HasBasisTransformations())) {
+        coverage = coverage->Intersection(coverage_limit.value());
+      }
```
Comment (jonahwilliams) [inline impeller/entity/entity_pass.cc:145]: nit: this code feels complicated enough that you might want to pull it into a helper method.

**Review by jonahwilliams — APPROVED** [review]
LGTM

File: impeller/aiks/paint.cc [hunk 4]
```diff
@@ -60,14 +60,21 @@ std::shared_ptr<Contents> Paint::WithFilters(
     std::shared_ptr<Contents> input) const {
   input = WithColorFilter(input, /*absorb_opacity=*/true);
```
Comment (chinmaygarde) [inline impeller/aiks/paint.cc:61]: Dry by comment on an unrelated item: Instead of booleans for absorb_opacity and is_subpass, I'd rather have a typed enum with AbsorbOpacity::kYes and AbsorbOpacity::kNo and the like.

File: impeller/entity/contents/filters/filter_contents.h (hunk 1 above)
Comment (bdero) [reply impeller/entity/contents/filters/filter_contents.h]: Done.

File: impeller/entity/contents/filters/filter_contents.h (hunk 2 above)
Comment (bdero) [reply impeller/entity/contents/filters/filter_contents.h]: Yeah, inverted this.

File: impeller/entity/entity_pass.cc (hunk 3 above)
Comment (bdero) [reply impeller/entity/entity_pass.cc]: Done.

File: impeller/aiks/paint.cc (hunk 4 above)
Comment (bdero) [reply impeller/aiks/paint.cc:61]: Added an issue about this: https://github.com/flutter/flutter/issues/134681

File: impeller/entity/entity_pass.cc [hunk 5]
```diff
@@ -135,6 +162,16 @@ std::optional<Rect> EntityPass::GetElementsCoverage(
...
     const EntityPass& subpass,
     std::optional<Rect> coverage_limit) const {
+  std::shared_ptr<FilterContents> image_filter =
+      subpass.delegate_->WithImageFilter(Rect(), subpass.xformation_);
+
+  // If the filter graph transforms the basis of the subpass, then its space
+  // has deviated too much from the parent pass to safely intersect with the
+  // pass coverage limit.
+  coverage_limit =
+      (image_filter && image_filter->IsTranslationOnly() ? std::nullopt
```
Comment (flar) [inline impeller/entity/entity_pass.cc:172]: Should this be `!IsTranslationOnly()`?

Comment (bdero) [reply impeller/entity/entity_pass.cc:172]: Yeah
